File: src/api/mock/fusionScore.ts
import type { RiskState } from '@/models/entities';

/**
 * Réimplémentation TypeScript fidèle du moteur de fusion — CDC_2_Serveur.docx §5.4.
 * Formule vérifiée textuellement dans le document (poids, bonus de concordance,
 * amplification contextuelle, seuils 30/70).
 */
export interface SubScoresInput {
  sUniversel: number; // 0..1
  sDeclaratif: number; // 0..1
  sGeo: number; // 0..1, avant atténuation par la confiance
  sMouvement: number; // 0..1
  confidence: number; // 0..100 — maturité du modèle (couche 3)
  contexteNuit: boolean;
  horsPerimetre: boolean;
}

export interface FusionResult {
  score: number; // 0..100
  sGeoEff: number;
  concordanceBonusApplied: boolean;
  contextAmplificationApplied: boolean;
}

const W_UNIV = 0.35;
const W_DECL = 0.3;
const W_GEO = 0.2;
const W_MOUV = 0.15;

export function computeFusionScore(input: SubScoresInput): FusionResult {
  const conf = Math.max(0, Math.min(100, input.confidence)) / 100;
  const sGeoEff = input.sGeo * conf;

  let base =
    W_UNIV * input.sUniversel + W_DECL * input.sDeclaratif + W_GEO * sGeoEff + W_MOUV * input.sMouvement;

  const activeSignals = [input.sUniversel, input.sDeclaratif, sGeoEff, input.sMouvement].filter(
    (x) => x >= 0.6
  ).length;
  const concordanceBonusApplied = activeSignals >= 2;
  if (concordanceBonusApplied) base = Math.min(1, base * 1.25);

  const contextAmplificationApplied = input.contexteNuit || input.horsPerimetre;
  if (contextAmplificationApplied) base = Math.min(1, base * 1.15);

  return {
    score: Math.round(base * 100),
    sGeoEff,
    concordanceBonusApplied,
    contextAmplificationApplied,
  };
}

export function stateFromScore(score: number): Exclude<RiskState, 'disparition'> {
  if (score >= 70) return 'urgence';
  if (score >= 30) return 'prealerte';
  return 'veille';
}

const RANK: Record<RiskState, number> = { veille: 0, prealerte: 1, urgence: 2, disparition: 3 };

/**
 * Hystérésis : au moins deux mesures consécutives doivent confirmer un changement
 * d'état, et une désescalade exige de redescendre nettement sous le seuil, pour
 * éviter le "clignotement" — CDC2 §5.4.
 */
export class HysteresisGate {
  private buffer: Exclude<RiskState, 'disparition'>[] = [];
  private confirmed: Exclude<RiskState, 'disparition'> = 'veille';

  constructor(initial: Exclude<RiskState, 'disparition'> = 'veille') {
    this.confirmed = initial;
  }

  get currentState() {
    return this.confirmed;
  }

  update(rawState: Exclude<RiskState, 'disparition'>, score: number): Exclude<RiskState, 'disparition'> {
    this.buffer.push(rawState);
    if (this.buffer.length > 2) this.buffer.shift();

    const escalation = RANK[rawState] > RANK[this.confirmed];
    const deescalation = RANK[rawState] < RANK[this.confirmed];
    const twoConsecutive = this.buffer.length === 2 && this.buffer[0] === this.buffer[1] && this.buffer[1] === rawState;

    if (escalation && twoConsecutive) {
      this.confirmed = rawState;
    } else if (deescalation && twoConsecutive) {
      const clearMargin =
        this.confirmed === 'urgence' ? score < 60 : this.confirmed === 'prealerte' ? score < 25 : true;
      if (clearMargin) this.confirmed = rawState;
    }
    return this.confirmed;
  }

  reset(state: Exclude<RiskState, 'disparition'> = 'veille') {
    this.confirmed = state;
    this.buffer = [];
  }
}

File: server/app/services/fusion_score.py
"""
Moteur de fusion du score de risque (section 5.4 du CDC).

Sous-scores normalisés 0-1 → fusion pondérée → score global 0-100.
"""

from dataclasses import dataclass, field
from typing import List, Optional


@dataclass
class SubScores:
    universel: float = 0.0
    declaratif: float = 0.0
    geo: float = 0.0
    mouvement: float = 0.0


@dataclass
class FusionResult:
    score: int
    state: str
    reasons: List[str]
    sub_scores: dict


# Poids de base (CDC section 5.4)
W_UNIV = 0.35
W_DECL = 0.30
W_GEO = 0.20
W_MOUV = 0.15

# Seuils (CDC section 5.4)
THRESHOLD_PREALERTE = 30
THRESHOLD_URGENCE = 70

# Seuil de concordance
CONCORDANCE_THRESHOLD = 0.6
NB_ACTIF_MIN = 2
CONCORDANCE_BONUS = 1.25

# Amplification contextuelle
CONTEXT_BONUS = 1.15


def compute_fusion(
    sub: SubScores,
    model_confidence: float = 1.0,
    contexte_nuit: bool = False,
    hors_perimetre: bool = False,
    previous_state: Optional[str] = None,
) -> FusionResult:
    reasons = []

    # Sous-score géographique atténué par la confiance du modèle
    s_geo_eff = sub.geo * model_confidence

    # Combinaison pondérée
    base = W_UNIV * sub.universel + W_DECL * sub.declaratif + W_GEO * s_geo_eff + W_MOUV * sub.mouvement

    # Règle de concordance
    actifs = sum(1 for x in [sub.universel, sub.declaratif, s_geo_eff, sub.mouvement] if x >= CONCORDANCE_THRESHOLD)
    if actifs >= NB_ACTIF_MIN:
        base = min(1.0, base * CONCORDANCE_BONUS)
        reasons.append(f"concordance:{actifs} signaux actifs")

    # Contexte
    if contexte_nuit:
        base = min(1.0, base * CONTEXT_BONUS)
        reasons.append("contexte:nuit")
    if hors_perimetre:
        base = min(1.0, base * CONTEXT_BONUS)
        reasons.append("contexte:hors_perimetre")

    score = round(base * 100)

    # Hystérésis : vérifier les mesures consécutives
    state = _determine_state(score, previous_state)

    return FusionResult(
        score=score,
        state=state,
        reasons=reasons,
        sub_scores={
            "universel": round(sub.universel * 100),
            "declaratif": round(sub.declaratif * 100),
            "geo": round(sub.geo * 100),
            "mouvement": round(sub.mouvement * 100),
        },
    )


def _determine_state(score: int, previous_state: Optional[str]) -> str:
    if previous_state == "disparition":
        return "disparition"

    if previous_state == "urgence":
        if score >= THRESHOLD_PREALERTE:
            return "urgence"
        return "veille"

    if previous_state == "prealerte":
        if score >= THRESHOLD_URGENCE:
            return "urgence"
        if score < THRESHOLD_PREALERTE:
            return "veille"
        return "prealerte"

    if score >= THRESHOLD_URGENCE:
        return "urgence"
    if score >= THRESHOLD_PREALERTE:
        return "prealerte"

    return "veille"
